No longer possible to use asterisk to see running executions across all projects

**Describe the bug**
Until recent versions it was possible to get details of running executions for all projects using the '*' symbol. 

```
curl -v -H X-Rundeck-Auth-Token:"${RD_TOKEN}" 'http://localhost:4440/api/35/project/*/executions/running'
* About to connect() to localhost port 4440 (#0)
*   Trying ::1...
* Connected to localhost (::1) port 4440 (#0)
> GET /api/35/project/*/executions/running HTTP/1.1
> User-Agent: curl/7.29.0
> Host: localhost:4440
> Accept: */*
> X-Rundeck-Auth-Token:REDACTED
> 
< HTTP/1.1 500 Server Error
< Set-Cookie: JSESSIONID=node01iz7f0hczn556l3denhp1vobb180037.node0; Path=/; HttpOnly
< Connection: close
< 
* Closing connection 0
```

This now returns a 500 error and outputs the following 

```
Caused by: java.lang.IllegalArgumentException: Project does not exist: *
        at rundeck.services.ProjectManagerService.$tt__getFrameworkProject(ProjectManagerService.groovy:118)
        at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
        at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
        at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
        at rundeck.services.FrameworkService.getFrameworkProject(FrameworkService.groovy:297)
        at rundeck.services.FrameworkService.getAuthContextForSubjectAndProject(FrameworkService.groovy:723)
        at rundeck.AuthTagLib$_closure5.doCall(AuthTagLib.groovy:234)
        at org.grails.taglib.TagOutput.captureTagOutput(TagOutput.java:78)
        at org.grails.taglib.TagLibraryMetaUtils$_registerMethodMissingForTags_closure3.doCall(TagLibraryMetaUtils.groovy:59)
        at gsp_rundeckapp_common_sidebar_gsp.run(gsp_rundeckapp_common_sidebar_gsp.groovy:151)
        at org.grails.gsp.GroovyPageWritable.doWriteTo(GroovyPageWritable.java:162)
        at org.grails.gsp.GroovyPageWritable.writeTo(GroovyPageWritable.java:82)
        at org.grails.web.gsp.GroovyPagesTemplateRenderer.makeTemplate(GroovyPagesTemplateRenderer.java:232)
        at org.grails.web.gsp.GroovyPagesTemplateRenderer.render(GroovyPagesTemplateRenderer.java:113)
        at org.grails.plugins.web.taglib.RenderTagLib$_closure7.doCall(RenderTagLib.groovy:367)
        at org.grails.gsp.GroovyPage.invokeTagLibClosure(GroovyPage.java:446)
        at org.grails.gsp.GroovyPage.invokeTag(GroovyPage.java:364)
        ... 98 common frames omitted
```
According to https://docs.rundeck.com/docs/api/rundeck-api.html#getting-executions-for-a-job I should be able to do this. It works fine if I specify a valid project name.

**My Rundeck detail**

* Rundeck version: 3.2.6 or 3.2.7
* install type: RPM
* OS Name/version: RHEL 7.7
* DB Type/version: postgres

**To Reproduce**
Steps to reproduce the behavior:
1. Make an API call via API as shown above
2. Note the 500 Error and associated error message as shown above

**Expected behavior**
I should be given back details on the overall number of running exections on a given node

**Additional context**

I also see this when I try to do the equivalent using the CLI (I note here that the regex should allow for the *) 

``` 
RD_PROJECT='*' rd executions list            
Input error for [rd executions list]: Cannot match (*) to pattern: /^[-_a-zA-Z0-9+][-\._a-zA-Z0-9+]*$/ : RD_PROJECT
You can use: "rd executions list -h" to get help.
```

Any help appreciated
Thanks in advance
-Padraig

--- a/rundeckapp/grails-app/controllers/rundeck/controllers/MenuController.groovy
+++ b/rundeckapp/grails-app/controllers/rundeck/controllers/MenuController.groovy
@@ -3165,40 +3165,70 @@ class MenuController extends ControllerBase implements ApplicationContextAware{
     /**
      * API: /executions/running, version 1
      */
-    def apiExecutionsRunning (){
-        if(!apiAuthorizedForEvent(params.project,[AuthConstants.ACTION_READ])){
+    def apiExecutionsRunning () {
+
+        //allow project='*' to indicate all projects
+        def allProjects = request.api_version >= ApiVersions.V9 && params.project == '*'
+
+        if (!allProjects && !apiAuthorizedForEvent(params.project, [AuthConstants.ACTION_READ])) {
             return
         }
         if (!apiService.requireApi(request, response)) {
             return
         }
-        if(!params.project){
+        if (!params.project) {
             return apiService.renderErrorFormat(response, [status: HttpServletResponse.SC_BAD_REQUEST,
-                    code: 'api.error.parameter.required', args: ['project']])
+                                                           code  : 'api.error.parameter.required', args: ['project']])
         }
         //test valid project
-
-        //allow project='*' to indicate all projects
-        def allProjects = request.api_version >= ApiVersions.V9 && params.project == '*'
-        if(!allProjects){
-            if(!apiService.requireExists(response,frameworkService.existsFrameworkProject(params.project),['project',params.project])){
+        if (!allProjects) {
+            if (!apiService.requireExists(response, frameworkService.existsFrameworkProject(params.project), ['project', params.project])) {
                 return
             }
         }
-        if (request.api_version < ApiVersions.V14 && !(response.format in ['all','xml'])) {
-            return apiService.renderErrorXml(response,[
-                    status:HttpServletResponse.SC_UNSUPPORTED_MEDIA_TYPE,
-                    code: 'api.error.item.unsupported-format',
-                    args: [response.format]
+        if (request.api_version < ApiVersions.V14 && !(response.format in ['all', 'xml'])) {
+            return apiService.renderErrorXml(response, [
+                    status: HttpServletResponse.SC_UNSUPPORTED_MEDIA_TYPE,
+                    code  : 'api.error.item.unsupported-format',
+                    args  : [response.format]
             ])
         }
 
-        QueueQuery query = new QueueQuery(runningFilter:'running',projFilter:params.project)
-        if(params.max){
-            query.max=params.int('max')
+        def projectNameAuthorized = "";
+
+        if (allProjects){
+            def authContext = frameworkService.getAuthContextForSubject(session.subject)
+            def projectNames = frameworkService.projectNames(authContext)
+
+            if (!projectNames || projectNames.isEmpty()) {
+                return apiService.renderErrorFormat(response, [status: HttpServletResponse.SC_UNAUTHORIZED,
+                                                               code  : 'api.error.execution.project.notfound', args: [params.project]])
+            }
+
+            for (names in projectNames) {
+                if (apiAuthorizedForEvent(names, [AuthConstants.ACTION_READ])) {
+                    if (projectNameAuthorized.isEmpty()) {
+                        projectNameAuthorized = names;
+                    } else {
+                        projectNameAuthorized = projectNameAuthorized + "," + names;
+                    }
+                }
+            }
+
+            if (!projectNameAuthorized || projectNameAuthorized.isEmpty()) {
+                return apiService.renderErrorFormat(response, [status: HttpServletResponse.SC_UNAUTHORIZED,
+                                                               code  : 'api.error.execution.project.notfound', args: [params.project]])
+            }
+        } else {
+            projectNameAuthorized = params.project
         }
-        if(params.offset){
-            query.offset=params.int('offset')
+
+        QueueQuery query = new QueueQuery(runningFilter: 'running', projFilter: projectNameAuthorized)
+        if (params.max) {
+            query.max = params.int('max')
+        }
+        if (params.offset) {
+            query.offset = params.int('offset')
         }
 
         if (request.api_version >= ApiVersions.V31 && params.jobIdFilter) {

--- a/rundeckapp/src/test/groovy/rundeck/controllers/MenuControllerSpec.groovy
+++ b/rundeckapp/src/test/groovy/rundeck/controllers/MenuControllerSpec.groovy
@@ -52,6 +52,7 @@ import rundeck.Workflow
 import rundeck.services.ApiService
 import rundeck.services.AuthorizationService
 import rundeck.services.ConfigurationService
+import rundeck.services.ExecutionService
 import rundeck.services.FrameworkService
 import rundeck.services.JobSchedulesService
 import rundeck.services.ScheduledExecutionService
@@ -1645,4 +1646,214 @@ class MenuControllerSpec extends HibernateSpec implements ControllerUnitTest<Men
         "test"                 | true           | null
         "test"                 | false          | "grouped"
     }
+
+    def "test list all projects with a valid project"() {
+        given:
+        controller.apiService = Mock(ApiService){
+            1 * requireApi(_,_) >> true
+        }
+
+        controller.userService = Mock(UserService){}
+
+        UserAndRolesAuthContext test = Mock(UserAndRolesAuthContext) {
+            getUsername() >> 'test'
+            getRoles() >> new HashSet<String>(['test'])
+        }
+
+        def projectMock = Mock(IRundeckProject) {
+            getProperties() >> [:]
+            getName() >> 'aProject'
+        }
+
+        def executionService = Mock(ExecutionService){
+            finishQueueQuery(_,_,_) >> [:]
+        }
+
+        controller.executionService = executionService
+
+        controller.frameworkService = Mock(FrameworkService){
+            getRundeckBase() >> ''
+            getFrameworkProject(_) >> projectMock
+            getServerUUID() >> 'uuid'
+            authorizeProjectResourceAll(_, AuthorizationUtil.resourceType('event'),['read'], 'aProject') >> true
+            1 * getAuthContextForSubject(_) >> test
+
+        }
+
+        params.project = '*'
+        request.api_version = 35
+        def action = 'read'
+
+        when:
+        def result = controller.apiExecutionsRunning()
+
+        then:
+        0 * controller.frameworkService.getAuthContextForSubjectAndProject(_, '*')
+        1 * controller.frameworkService.projectNames(_) >> ['aProject']
+        0 * controller.apiService.renderErrorFormat(_,_)
+        1 * controller.executionService.queryQueue(_) >> {
+            assert it[0].projFilter == 'aProject'
+
+        }
+    }
+
+    def "test list all projects with an invalid project"() {
+        given:
+        controller.apiService = Mock(ApiService){
+            1 * requireApi(_,_) >> true
+        }
+
+        controller.userService = Mock(UserService){}
+
+        UserAndRolesAuthContext test = Mock(UserAndRolesAuthContext) {
+            getUsername() >> 'test'
+            getRoles() >> new HashSet<String>(['test'])
+        }
+
+        def projectMock = Mock(IRundeckProject) {
+            getProperties() >> [:]
+            getName() >> 'aProject'
+        }
+
+        def executionService = Mock(ExecutionService){
+            finishQueueQuery(_,_,_) >> [:]
+        }
+
+        controller.executionService = executionService
+
+        controller.frameworkService = Mock(FrameworkService){
+            getRundeckBase() >> ''
+            getFrameworkProject(_) >> projectMock
+            getServerUUID() >> 'uuid'
+            authorizeProjectResourceAll(_, AuthorizationUtil.resourceType('event'),['read'], 'aProject') >> false
+            1 * getAuthContextForSubject(_) >> test
+
+        }
+
+        params.project = '*'
+        request.api_version = 35
+        def action = 'read'
+
+        when:
+        def result = controller.apiExecutionsRunning()
+
+        then:
+        0 * controller.frameworkService.getAuthContextForSubjectAndProject(_, '*')
+        1 * controller.frameworkService.projectNames(_) >> ['aProject']
+        1 * controller.apiService.renderErrorFormat(_,{map->
+            map.status==401 && map.code=='api.error.execution.project.notfound'
+        })
+        0 * controller.executionService.queryQueue(_) >> {
+            assert it[0].projFilter == 'aProject'
+
+        }
+
+    }
+
+    def "test list all projects with an invalid and a valid project"() {
+        given:
+        controller.apiService = Mock(ApiService){
+            1 * requireApi(_,_) >> true
+        }
+
+        controller.userService = Mock(UserService){}
+
+        UserAndRolesAuthContext test = Mock(UserAndRolesAuthContext) {
+            getUsername() >> 'test'
+            getRoles() >> new HashSet<String>(['test'])
+        }
+
+        def projectMock = Mock(IRundeckProject) {
+            getProperties() >> [:]
+            getName() >> ['aProject', 'bProject']
+        }
+
+        def executionService = Mock(ExecutionService){
+            finishQueueQuery(_,_,_) >> [:]
+        }
+
+        controller.executionService = executionService
+
+        controller.frameworkService = Mock(FrameworkService){
+            getRundeckBase() >> ''
+            getFrameworkProject(_) >> projectMock
+            getServerUUID() >> 'uuid'
+            authorizeProjectResourceAll(_, AuthorizationUtil.resourceType('event'),['read'], 'aProject') >> true
+            1 * getAuthContextForSubject(_) >> test
+
+        }
+
+        params.project = '*'
+        request.api_version = 35
+        def action = 'read'
+
+        when:
+        def result = controller.apiExecutionsRunning()
+
+        then:
+        0 * controller.frameworkService.getAuthContextForSubjectAndProject(_, '*')
+        1 * controller.frameworkService.projectNames(_) >> ['aProject', 'bProject']
+        0 * controller.apiService.renderErrorFormat(_,{map->
+            map.status==401 && map.code=='api.error.execution.project.notfound'
+        })
+        1 * controller.executionService.queryQueue(_) >> {
+            assert it[0].projFilter == 'aProject'
+
+        }
+
+    }
+
+    def "test list all projects with multiple valid projects"() {
+        given:
+        controller.apiService = Mock(ApiService){
+            1 * requireApi(_,_) >> true
+        }
+
+        controller.userService = Mock(UserService){}
+
+        UserAndRolesAuthContext test = Mock(UserAndRolesAuthContext) {
+            getUsername() >> 'test'
+            getRoles() >> new HashSet<String>(['test'])
+        }
+
+        def projectMock = Mock(IRundeckProject) {
+            getProperties() >> [:]
+            getName() >>  ['aProject', 'bProject', 'cProject']
+        }
+
+        def executionService = Mock(ExecutionService){
+            finishQueueQuery(_,_,_) >> [:]
+        }
+
+        controller.executionService = executionService
+
+        controller.frameworkService = Mock(FrameworkService){
+            getRundeckBase() >> ''
+            getFrameworkProject(_) >> projectMock
+            getServerUUID() >> 'uuid'
+            authorizeProjectResourceAll(_, AuthorizationUtil.resourceType('event'),['read'], _) >> {
+                it[3] == 'aProject' ||  it[3] == 'bProject'||  it[3] == 'cProject'
+            }
+
+            1 * getAuthContextForSubject(_) >> test
+
+        }
+
+        params.project = '*'
+        request.api_version = 35
+        def action = 'read'
+
+        when:
+        def result = controller.apiExecutionsRunning()
+
+        then:
+        0 * controller.frameworkService.getAuthContextForSubjectAndProject(_, '*')
+        1 * controller.frameworkService.projectNames(_) >> ['aProject', 'bProject', 'cProject' ]
+        0 * controller.apiService.renderErrorFormat(_,_)
+        1 * controller.executionService.queryQueue(_) >> {
+            assert it[0].projFilter == 'aProject,bProject,cProject'
+
+        }
+    }
+
 }

